services/worker-manager/src/providers/azure.js

```diff
@@ -707,47 +713,60 @@ class AzureProvider extends Provider {
-                // if the VM has ever been in a failing power state
-                _.some(powerStates, v => failPowerStates.has(v))
-      ) {
-        state = await this.removeWorker({
-          worker,
-          reason: `failed state; provisioningState=${provisioningState}, powerStates=${powerStates.join(', ')}`,
-        });
-      } else {
-        const { workerPoolId } = worker;
-        const workerPool = await WorkerPool.get(this.db, workerPoolId);
-        if (workerPool) {
-          await this.reportError({
-            workerPool,
-            kind: 'creation-error',
-            title: 'Encountered unknown VM provisioningState or powerStates',
-            description: `Unknown provisioningState ${provisioningState} or powerStates: ${powerStates.join(', ')}`,
-          });
+
+      // count this worker as having been seen for later logging
+      this.seen[worker.workerPoolId] += worker.capacity || 1;
+
+      // See https://docs.microsoft.com/en-us/azure/virtual-machines/states-lifecycle
+      // for background on the VM lifecycle.
+      let isFailed = false;
+      let reason;
+
+      if (_.some(powerStates, state => failPowerStates.has(state))) {
+        // A VM can transition to one of failPowerStates through an Azure issue of some sort, by being
+        // manually terminated (e.g., in the web UI), or by being halted from within the VM.  In this
+        // case, we consider the worker failed and begin to remove it.
+        isFailed = true;
+        reason = `failed power state; powerStates=${powerStates.join(', ')}`;
+      }
+
+      if (!isFailed && state === states.REQUESTED) {
+        // It's possible for a newly-requested VM to be running (PowerState/running), but have failed
+        // provisioning.  In this case the VM isn't doing any work, but billing continues.  So, we want
+        // to catch this case and also consider it failed.  These state codes have the form
+        // `ProvisioningState/failed/<SomeCode>`.
+        let failedProvisioningCodes = powerStates
+          .filter(state => state.startsWith('ProvisioningState/failed/'))
+          .map(state => state.split('/')[2]);
+
+        // any failed-provisioning code is treated as a failure
+        if (failedProvisioningCodes.length > 0) {
+          isFailed = true;
+          reason = `failed provisioning power state; powerStates=${powerStates.join(', ')}`;
         }
       }
+
+      if (!isFailed && worker.providerData.terminateAfter && worker.providerData.terminateAfter < Date.now()) {
+        // If the worker has not checked in recently enough, we consider it failed regardless of the Azure lifecycle
+        isFailed = true;
+        reason = 'terminateAfter time exceeded';
+      }
+
+      if (isFailed) {
+        // On failure, call `removeWorker`, which works to remove the resources associated with the VM
+        // and eventually mark it as STOPPED
+        state = await this.removeWorker({ worker, reason });
+      }
```

Comment:
Is it possible that a piece of code ends up going in multiple of these if statements? Right now, the `reason` will always reflect the last matched condition. I wonder if we should rather be using `else if` instead. Thoughts?

Reply:
Each is preceded by `!isFailed &&` which aims to stick with the *first* matched condition.  Using else-if is hard with nested if's:
```
if (..) {
  ..
  if (..) {
    isFailed = true;
  }
} else if (..) {
}
```
isn't quite what we want.

Reply:
Also, come to think of it, the ordering of the checks isn't that important -- if it's both later than terminateAfter *and* the VM is in some bad state, then a removeWorker call with either reason would be fine.